windows/mdns_impl.c

```diff
@@ -0,0 +1,525 @@
+          first_ipv4 = 0;
+          log_addr = 1;
+        }
+        has_ipv4 = 1;
+        if (num_sockets < max_sockets) {
+          saddr->sin_port = htons(port);
+          int sock = mdns_socket_open_ipv4(saddr);
+          if (sock >= 0) {
+            sockets[num_sockets++] = sock;
+            log_addr = 1;
+          }
+          else {
+            log_addr = 0;
+          }
+        }
+        if (log_addr) {
+          char buffer[128];
+          mdns_string_t addr = ipv4_address_to_string(buffer, sizeof(buffer), saddr,
+            sizeof(struct sockaddr_in));
+          printf("Local IPv4 address: %.*s\n", MDNS_STRING_FORMAT(addr));
+        }
+      }
+    }
+    else if (ifa->ifa_addr->sa_family == AF_INET6) {
+      struct sockaddr_in6* saddr = (struct sockaddr_in6*)ifa->ifa_addr;
+      static const unsigned char localhost[] = { 0, 0, 0, 0, 0, 0, 0, 0,
+                                                0, 0, 0, 0, 0, 0, 0, 1 };
+      static const unsigned char localhost_mapped[] = { 0, 0, 0,    0,    0,    0, 0, 0,
+                                                       0, 0, 0xff, 0xff, 0x7f, 0, 0, 1 };
+      if (memcmp(saddr->sin6_addr.s6_addr, localhost, 16) &&
+        memcmp(saddr->sin6_addr.s6_addr, localhost_mapped, 16)) {
+        int log_addr = 0;
+        if (first_ipv6) {
+          service_address_ipv6 = *saddr;
+          first_ipv6 = 0;
+          log_addr = 1;
+        }
+        has_ipv6 = 1;
+        if (num_sockets < max_sockets) {
+          saddr->sin6_port = htons(port);
+          int sock = mdns_socket_open_ipv6(saddr);
+          if (sock >= 0) {
+            sockets[num_sockets++] = sock;
+            log_addr = 1;
+          }
+          else {
+            log_addr = 0;
+          }
+        }
+        if (log_addr) {
+          char buffer[128];
+          mdns_string_t addr = ipv6_address_to_string(buffer, sizeof(buffer), saddr,
+            sizeof(struct sockaddr_in6));
+          printf("Local IPv6 address: %.*s\n", MDNS_STRING_FORMAT(addr));
+        }
+      }
+    }
+  }
+
+  freeifaddrs(ifaddr);
+
+#endif
+
+  return num_sockets;
+}
+
+
+
+// Open sockets to listen to incoming mDNS queries on port 5353
+static int
+open_service_sockets(int* sockets, int max_sockets) {
+  // When recieving, each socket can recieve data from all network interfaces
+  // Thus we only need to open one socket for each address family
+  int num_sockets = 0;
+
+  // Call the client socket function to enumerate and get local addresses,
+  // but not open the actual sockets
+  open_client_sockets(0, 0, 0);
+
+  if (num_sockets < max_sockets) {
+    struct sockaddr_in sock_addr;
+    memset(&sock_addr, 0, sizeof(struct sockaddr_in));
+    sock_addr.sin_family = AF_INET;
+#ifdef _WIN32
+    sock_addr.sin_addr = in4addr_any;
+#else
+    sock_addr.sin_addr.s_addr = INADDR_ANY;
+#endif
+    sock_addr.sin_port = htons(MDNS_PORT);
+#ifdef __APPLE__
+    sock_addr.sin_len = sizeof(struct sockaddr_in);
+#endif
```

Comment:
This file has many compiler directives for other platforms than Windows. I think it would be good to remove what we don't need.